Gs 00004 (#32)
bitsandvolts/bitsandvolts.github.io: _posts/2021-10-01-gs_00004_basic_shell_script.md

---
layout: single

> title: Basic Shell Script

Reviewer: Would "Scripts" sound better?

category: Getting Started
permalink: /categories/getting_started/basic_shell_script
excerpt: "Automate repeating tasks"
author: Bjoern Hartmann
toc: true
toc_label: "Content"
toc_sticky: true
toc_icon: "align-justify"
sidebar:
  nav: "getting_started_0"
header:
  overlay_color: "#000"
  overlay_filter: "0.5"
  overlay_image: /assets/images/getting_started.jpg
---

<div class='note-normal'>
<top4>Prerequisites</top4>
<ul>
    <li><a href="/categories/getting_started/terminal_basics">Terminal Basics I</a></li>
    <li><a href="/categories/getting_started/terminal_basics2">Terminal Basics II</a></li>
</ul>
</div>

> We are glad you made it back to our next post. Is your head ready to learn a

Reviewer: Would sound more "exciting" ^^
```suggestion
We are glad you made it back to our next post! Is your head ready to learn a
```

That "is your head ready" sounds a bit odd to me. But you can leave it, if you think it's ok for you.

> very cool thing called **Shell Script**? Let's go...

Reviewer: Better use plural here (scripts).

In our very first post, we stated that using a terminal has many advantages over
a graphical interface. To convince you, we introduced very basic but powerful
commands that you can use to control your computer with a terminal. If you have
no idea what a terminal is or if you forgot some commands, just use one of the

> two links above to go back to the previous posts. Even things like the

Reviewer: Did you mean "Eventhough" here?
```suggestion
two links above to go back to the previous posts. Eventhough things like the
```

Author: Even though, right?

Reviewer: You are right

autocompletion and the "\*" operator are reason enough for us to use the
terminal, this might not convince everyone. So, we want to go one step further.
Are you ready?

<div style="overflow: auto;" markdown="1">
<figure class="img-right">
  <img src="/assets/images/gs_00004_josephine.jpg" alt="">
</figure>

Let's do some time traveling into the year 1886. There was a woman called
Josephine. She loved to celebrate parties. There was just one issue with that.
She hated to clean the dishes afterward. So, she tried to build a machine to do
it automatically for her. Finally, she was the first person in the world to
construct a working dishwasher.

Even though we do not expect you to invent something so meaningful as a
dishwasher, the idea behind the concept should come to your mind very often.

> Whenever you have to do something regularly you should ask yourself if it is

Reviewer: To more emphasis on it:
```suggestion
Whenever you have to do something _regularly_ you should ask yourself if it is
```

> possible to automate it. This is more or less what a **Script** does. It calls a

Reviewer: ```suggestion
possible to _automate_ it. This is more or less what a **script** does. It calls a
```
Btw, I see it often that you write terms like "shell", "terminal" with a capital letter. I think it would be better to just keep them lower-case. You've made them bold anyways.

sequence of commands for you. We want to give you two examples that might be

> useful in the real world. Before that, we have to clarify the term **Shell**.

Reviewer: ```suggestion
useful in the real world. Before that, we have to clarify the term **shell**.
```

</div>

# Shell, Terminal, Console ...

All these terms are very confusing for beginners as well as advanced engineers.
Regularly, they are used interchangeably. Let's try to make them as clear as
possible. In the very beginning, computers were huge and usually had no user
interface. You needed some extra **physical device** that was connected to the

> computer, which was called **Console**. In contrast, the **Terminal** is not

Reviewer: ```suggestion
computer, which was called **console**. In contrast, the **terminal** is not
```

> something physical. It is simply a text input and output environment. So far so

Reviewer: If the terminal is not physical, where is this environment located?

```suggestion
something physical. It is simply a text input and output environment. So far, so
```

> good. You have a physical device (Console) with a keyboard or something similar

Reviewer: ```suggestion
good. You have a physical device (console) with a keyboard or something similar,
```
(use comma before "and")

> and kind of a program where you can type some text in (Terminal). But there is

Reviewer: ```suggestion
and kind of a program where you can type some text in (terminal). But there is
```

one step more. Something needs to read the text you are typing and **interpret**

> its meaning. The software that is doing this task is called **Shell**. If you

Reviewer: ```suggestion
its meaning. The software that is doing this task is called **shell**. If you
```

> enter a command like "ls", the shell _understands_ that you want to get the

Reviewer: ```suggestion
enter a command like `ls`, the shell _understands_ that you want to get the
```

content of the current folder. It executes the command and returns the output to

> the terminal.

Reviewer: I think the transition to modern computers is missing here.

Something that popped into my mind while reading this section: This whole topic shell/console/terminal might be a bit confusing for a total newbie. You could address this more in depth in a separate "bonus" post. And in the current post, simply explain that a shell is the program that reads your commands and runs them. I think, for the topic of this post, that would be sufficient.

Author: I moved it as bonus chapter to the end

As you can have different physical devices, you can also have different
terminals and shells. A different terminal might look different from another
one. A different shell might interpret different commands. Wait ... What? Yeah,
you heard correctly! Commands that work in one shell might work differently or
not at all in another. But don't worry! Just have it in mind for now! By the
way, the shell we used in the previous posts is called **Bash**. We might talk
about other shells in future posts, but for now, we stay with this commonly used
shell.

# Your first Bash Script

> Every morning the same procedure. You go to work, get your first coffee and then

Reviewer: Does every reader have this routine? Maybe say something like "the usual morning of a software developer looks like ..."

Author: recheck

open all required software for your job: Email client, meeting software,
browser, music app, and so on. Probably it might be different depending on your
daily work schedule. Wouldn't it be cool that at least the software opening part

> is done automatically? Let's take it as an example for the first script and open

Reviewer: ```suggestion
is done automatically? Let's take it as an example for the first script and open
```
That "and open a browser" could sound like a task for the reader.

Author: recheck

a browser. You already know how to do it with a terminal command, don't you? To
write your first script, open a new file:

    $ nano open_firefox.sh

<div class='note-normal' markdown="1">
_Note: '.sh' is the typical file extension for shell scripts_.
</div>

Enter the following text:

    #!/bin/bash

    # Print something to the terminal
    echo "Greetings from Bits&Volts (bitsandvolts.org)"

    # Open firefox
    firefox &

Press CTRL-X to close and 'y' (yes) to save the file. Before we run it, let's

> talk about it. The first line starting with the '#!' tells by what the script

Reviewer: Did you mean this:
```suggestion
talk about it. The first line starting with the '#!' tells by what shell the script
```

Author: does not need to be shell. recheck

Reviewer: Yes, sheabang could be anything. Maybe you can add a footnote or something, that in our case we use the sheabang to select a shell interpreter.

But decide for yourself; you can also ignore my suggestion.

Author: did you check the newest version?

shall be interpreted. In our case, we want to use **Bash**. What does the 'bin'
mean? The **bin** folder is a common folder where software programs such as the
bash shell are installed. So, '/bin/bash/' is nothing but the absolute path of
the shell program. The third line starting with '#' is just a comment that is
not interpreted. If you write complex scripts, it is helpful to put some

> commands, so that someone else can better understand it. 'echo' is a command

Reviewer: ```suggestion
comments, so that someone else can better understand it. 'echo' is a command
```

that we did not cover yet. It simply outputs some text. Then, we have another

> comment and the command to open firefox you already know. Note that we have to

Reviewer: ```suggestion
comment, and the command to open firefox you already know. Note that we have to
```

> use the '&' here, as explained in our last post. Try to run the script as you

Reviewer: Maybe add something like this as a reminder:
```suggestion
use the '&' here, as explained in our last post (it means that the program will be started in the background).

Try to run the script as you
```
Also, might be visually helpful to start a new paragraph.

would run a program:

    $ open_firefox.sh

> This will not work as the command cannot be found. To understand why we have to

Reviewer: ```suggestion
This will not work as the command cannot be found. To understand why, we have to
```

look a bit deeper into what happens if you run a program. Clearly, the program
has to be stored somewhere. But how can your operating system know where? The
answer is: It can't! It only can look into all your files and folders and search
for them. However, you can imagine that this can take a while, especially for
large systems. This is why there are predefined folders where the programs have

> to be. As we mentioned earlier, that '/bin' is, per default, such a folder. You

Reviewer: ```suggestion
to be. As we mentioned earlier, that '/bin' is, per default, such a folder. In general, there are three options for scripts:
```

have three options:

1. Move the script to a folder where it can be found. We do not want to do it
   here as it is not always practical in real-life situations.
2. Add your current folder to the so-called **PATH** variable. We explain in a
   later post how to do it. It is also not always a good solution. If you add
   too many folders it will take long to execute a command.
3. Run the program by its absolute path. This is always an easy solution.
   ```
   $ ./open_firefox.sh
   ```

> 

Reviewer: I would add here that we opted for option 3.

> Ups... Still does not work. But how can it be that the permission is denied? You

Reviewer: Maybe copy the error text here, otherwise the reader might get confused when he reads "permission denied" (not all readers might be running the commands themselves)

Author: recheck

created the file yourself in a folder that belongs to your home folder. What can

> be missing? We talk about access rights more in more detail in our next posts.

Reviewer: ```suggestion
be missing? We talk about access rights in more detail in our next posts.
```

> For now, we just tell you that you have the right to read and write the file.

Reviewer: I would use "permission" here:
```suggestion
For now, we just tell you that you have the permission to read and write the file.
```

> However, you need the right to execute the file. We can add it using:

Reviewer: ```suggestion
However, to run your script, you need the permission to _execute_ the file. We can add it using:
```

Imagine you are working in a company with multiple projects. We often experience
that every employee organizes the projects he or she is responsible for in
his/her own way. This, however, has a huge drawback. If there is a new
colleague, he/she has to find his/her own way. Moreover, if you take over the
project from a colleague you have to spend time to understand his/her project
organization. A better way would be to have a project template that everyone
uses. Let's create a very basic one!

First, you usually would create a folder with the project name. Then, inside
this folder, you might have a folder for the project sources and one for the
documentation. Additionally, there might be a **README** and a **LICENSE** file.
In the REAMDE you usually put basic information. For example, how you can
install the software or how your project is organized. The LICENSE file is
important if you want to publish your project. It tells others how they are
allowed to use it.

<div class='note-exercise' markdown="1">
Exercise:

Check our [GitHub repository](https://github.com/bitsandvolts/getting_started)
and find an example for a README and a LICENSE file.

</div>

> From the previous posts and the first bash script above you should already know

Reviewer: ```suggestion
From the previous posts, and the first bash script above you should already know
```

> how to do it. Use nano to open a file 'project_template.sh' and enter the

Reviewer: ```suggestion
how to create the necessary folders and files. Use `nano` to open a file 'project_template.sh' and enter the
```

following:

    #!/bin/bash

    # Print something to the terminal
    echo "Greetings from Bits&Volts (bitsandvolts.org)"

    # Create the project folder
    mkdir my_project

    # move into the project folder
    cd my_project

    # create other folders and files
    mkdir sources
    mkdir documentation
    touch README.md
    touch LICENSE

<div class='note-normal' markdown="1">
_Note: '.md' is a Markdown extension, which is typical for a README. We will talk about it in a future post._
</div>

Save the file, add the correct privileges and run it:

    $ chmod +x project_template.sh
    $ ./project_template.sh

If you list your directory contents, you should see the created project:

>     $ ls

Reviewer: I would also copy the output of `ls` here.

Author: hmm. but that depends on what is next to the script.... I think it is enough here

Reviewer: Oh, I initially thought the `project_template.sh` script would move into the project. So then ls would made sense. But it's actually not sourced, so yeah, don't change then.

> So far so good! However, not all projects are called 'my_project' and it would

Reviewer: ```suggestion
So far, so good! However, not all projects are called 'my_project' and it would
```
Actually, I checked online, and there people also using "so far so good" without comma. So decide by yourself what is better (I also made a "correction" of this somewhere above)

> be inconvenient to always change the script. That's why we want to introduce you

Reviewer: > "However, not all projects are called ‘my_project’. So, if you want to create a new project with a different name, you would have to manually edit the script. Doing this for every new project is very inconvenient."

This would be a nice bridge between what you have right now ("_not all projects are called ‘my_project_’" and "_would be inconvenient_")

Author: I do not understand what you want me to do here

Reviewer: Maybe this is simpler:
```suggestion
be inconvenient to always change the script, whenever you want to create a new project. That's why we want to introduce you
```
My point was that some newbies wouldn't be aware of the fact that they would have to manually edit the script for each new project. Maybe I am overthinking this, so what do you think?

to something called **arguments**. If you call your script you can give it
additional parameters that are used by the script. In particular, we want to
call our script like this:

>     $ ./project_template.sh project_name

Reviewer: I don't remember exactly, but did you explain that arguments are basically the words that come after the command name?

Author: recheck

Before we can do it, we have to make a tiny change. Within the script, such
arguments can be accessed with a
<span>$</span>-sign and a number. <span>$</span>1 would be the first argument,
<span>$</span>2 the second, and so on. In our script, we need to change the

> following lines:

Reviewer: For clarity:
```suggestion
following lines (the other lines can be left as they are):
```

    # Create the project folder
    mkdir $1

    # move into the project folder
    cd $1

Easy, isn't it? Now try to call the script

    $ ./project_template.sh project_name

> 

Reviewer: Maybe say here that there should be a new project folder called "project_name".

<div class='note-exercise' markdown="1">
Exercise:

> <span>$</span>0 holds the name of the script. Add a line to the script printing

Reviewer: Maybe explain that `$0` is in general the name/path of the program ?

Author: recheck

the script name using echo and <span>$</span>0.
Run the script using:

    $ ./project_template.sh hello

</div>

<div class='note-normal' markdown="1">

> _Note: We always use different project names whenever we run the script. What should happen if you create a project that already exists? Probably, the user should get a warning and the creation should be aborted. We will avoid this situation for now and cover such a check in a future post. Also, we should check that the user really entered an argument. Calling the script with no or more than one argument should fail as well._

Reviewer: ```suggestion
_Note: We always use different project names whenever we run the script. What would happen if you create a project that already exists? Probably, the user should get a warning and the creation should be aborted. We will avoid this situation for now and cover such a check in a future post. Also, we should check that the user really entered an argument. Calling the script with no or more than one argument should fail as well._
```

</div>

There is one last, but very important aspect. Inside the script, we used a 'cd'

> command, so we should be inside the project folder. However, as you used 'ls'

Reviewer: ```suggestion
command, so we should be inside the project folder after our script has run. However, as you used 'ls'
```

before, you might have noticed that this is not the case. You are still where

> you were before you started the script. The reason is the following. If you run

Reviewer: ```suggestion
you were before you started the script. The reason is the following: If you run
```

> the script it is not run in your current shell. Internally, a new shell is

Reviewer: For emphasis:
```suggestion
the script it is not run in your _current_ shell. Internally, a _new_ shell is
```

> created which is closed at the end of the script. So, the 'cd' has no impact on

Reviewer: Not sure if "created" is correct, what do you think?
```suggestion
opened which executes the commands in our script. The shell is then automatically closed at the end of the script. So, the 'cd' has no impact on
```

> your current shell. Sometimes, however, you want to change it. For your project

Reviewer: ```suggestion
your current shell. Sometimes, however, you want to change you current working directory automatically. For your project
```

template, it might be useful to directly move into the project folder. If you
want to do so, you have to **source** the script:

    $ source ./project_template.sh i_am_a_project

> 

Reviewer: Maybe explain what `source` does: the current shell itself will be used to execute the script.

Verify that you moved inside the project folder:

    $ pwd

# Summary

> 

Reviewer: Maybe add a short introduction to this section here.

Author: like this?

Reviewer: Weird, I don't see any change. Also for other updates you did. Could you check again?

Author: You asked for an introduction to the summary, right? It's starting with "Yaay! Your first script are running ..."

> You can find the two example script in our

Reviewer: ```suggestion
You can find the two example scripts in our
```

[GitHub repository](https://github.com/bitsandvolts/getting_started) of the
"Getting Started" category. We put a lot of work and energy into our blog posts
and the additional material, so please respect our
[terms of usage](https://github.com/bitsandvolts/getting_started/blob/master/LICENSE).

- A **script** can execute multiple commands for you

> - A **Console** is a physical device that let you communicate with a computer

Reviewer: Probably better to use lowercase to be more consistent?
```suggestion
- A **console** is a physical device that lets you communicate with a computer
```

> - The **Terminal** is a text input-output environment

Reviewer: ```suggestion
- A **terminal** is a text input-output environment
```

> - A **Shell** interprets your commands

Reviewer: ```suggestion
- A **shell** interprets your commands
```

- **Bash** is a commonly-used shell

> - At the beginning of a script, you usually tell by what it should be

Reviewer: What do you mean "by what"?

Author: better?

>   interpreted (#!/bin/bash)

Reviewer: ```suggestion
  interpreted (`#!/bin/bash`)
```

- Programs have to be located in defined folders to be found
- As an alternative, you can run it by its **absolute path**

> - A script needs the execution **privilege** (chmod +x name)

Reviewer: ```suggestion
- A script needs the execution **privilege** (`chmod +x name`)
```

> - Script **arguments** can be accessed by $1, $2, ...

Reviewer: ```suggestion
- Script **arguments** can be accessed by `$1`, `$2`, ...
```

> - Source a script to let it take effect on your current shell

Reviewer: ```suggestion
- **Source** a script to let it take effect on your current shell
```

> In the next post, we will talk about **privileges** and so-called **Environment

Reviewer: ```suggestion
In the next post, we will talk about **privileges** and so-called **environment
```

have to use the '&' here, as explained in our last post (it means that the
program will be started in the background).

Try to run the script as you would run a program:

    $ open_firefox.sh

This will not work as the command cannot be found. To understand why, we have to
look a bit deeper into what happens if you run a program. Clearly, the program
has to be stored somewhere. But how can your operating system know where? The
answer is: It can't! It only can look into all your files and folders and search
for them. However, you can imagine that this can take a while, especially for
large systems. This is why there are predefined folders where the programs have
to be. As we mentioned earlier, that '/bin' is, per default, such a folder. In
general, there are three options for scripts:

1. Move the script to a folder where it can be found. We do not want to do it
   here as it is not always practical in real-life situations.
2. Add your current folder to the so-called **PATH** variable. We explain in a
   later post how to do it. It is also not always a good solution. If you add
   too many folders it will take long to execute a command.

> 3. Run the program by its absolute path. This is always an easy solution and we

Reviewer: Maybe move the part saying that we want to use option 3 to a new paragraph. That way the list would stay as general as possible.

Author: recheck

    $ ls

So far, so good! However, not all projects are called 'my_project' and it would
be inconvenient to always change the script. That's why we want to introduce you
to something called **arguments**. If you call your script you can give it
additional parameters (arguments) that are used by the script. In particular, we
want to call our script like this:

    $ ./project_template.sh project_name

Before we can do it, we have to make a tiny change. Within the script, such
arguments can be accessed with a
<span>$</span>-sign and a number. <span>$</span>1 would be the first argument,
<span>$</span>2 the second, and so on. In our script, we need to change the
following lines (the other lines can be left as they are):

    # Create the project folder
    mkdir $1

    # move into the project folder
    cd $1

Easy, isn't it? Now try to call the script and a new folder called

> 'project_name' should apper:

Reviewer: "appear"

# Summary

Yaaay! Your first scripts are running. Doesn't that feel cool? Let's try to
summarize what we learned in this post. As always, you can find the two example
scripts in our
[GitHub repository](https://github.com/bitsandvolts/getting_started) of the
"Getting Started" category. We put a lot of work and energy into our blog posts
and the additional material, so please respect our
[terms of usage](https://github.com/bitsandvolts/getting_started/blob/master/LICENSE).

- A **script** can execute multiple commands for you
- A **shell** interprets your commands
- **Bash** is a commonly-used shell
- At the beginning of a script, you usually tell what program shall interpret
  the script ('#!/bin/bash')
- Programs have to be located in defined folders to be found
- As an alternative, you can run it by its **absolute path**
- A script needs the execution **privilege** ('chmod +x name')

> - Script **arguments** can be accessed by '<span>$</span>1', '<span>$</span>2',

Reviewer: Did you add the `span`? I think just dollar sign + number surrounded by **backticks**, ie inline code, would also do the job.

Same for the "chmod...".

  ...
- **Source** a script to let it take effect on your current shell

In the next post, we will talk about **privileges** and so-called **environment
variables**! Stay tuned, and don't forget to follow us on
[Twitter](https://twitter.com/bitsandvolts)! If you cannot wait to learn more,
be our guest in the following bonus section.

# Shell, Terminal, Console ... (Bonus)

> All these terms are very confusing for beginners as well as advanced engineers.

Reviewer: I would add a small introductory sentence. Maybe something like
> In this post we have introduced you to the shell. Sometimes though, you might hear terms like "terminal", or "console". ...
